Refactoring  branch

aleksandrzajbel25182 opened the pull request.

aleksandrzajbel25182 committed:
Create package stdin

aleksandrzajbel25182 committed:
Created the game class without implementation

aleksandrzajbel25182 committed:
creating an enumeration for the gallows

aleksandrzajbel25182 committed:
Create class Gallows , final string array

aleksandrzajbel25182 committed:
Create class Game, Hangman, StdoutGallows . Delete enum StageGallows.java

aleksandrzajbel25182 committed:
Удалил лишнии классы. Абстрактный класс Hangma
 переделал в интерфейс. Нужно сделать вывод сообщения на печать

aleksandrzajbel25182 committed:
Более менее работаю версия. Есть ошибки если проиграл необходимо сделать старт. Переделал интерфейс Answer имеет лишь один мето принимающий String. вывод в консоль происходит в классе Game от этого уйти не получилось.

aleksandrzajbel25182 committed:
Удалил лишний класс GameContext.java и поправил код на пробелы лишние

aleksandrzajbel25182 committed:
Переделал класс WordChoice.java в WordSource.java положил это в пакет Source. Так же метод чтение из файла вынес в интерфейс InputSource, реализация которого происходит в WordSource.
Класс WordSource.java теперь не запускает чтение из конструктора, а принимает лишь строку для подключения. В данном случае  - файл формата "text.txt"

aleksandrzajbel25182 committed:
схватил ошибку NullPointerException из- за того что не правильно написал название документа. Так же записал делать чтение файла, собственно добавил данную строчку

aleksandrzajbel25182 committed:
Press 'S' Start game

shining-mind commented on src/main/java/Application.java:
```diff
@@ -1,6 +1,33 @@
+import Source.WordSource;
+import stdin.StdinCharSource;
+import stdout.AttemptsHangman;
```
Неочевидно, почему AttemptsHangman расположен в namespace stdout.

shining-mind commented on src/main/java/Application.java:
```diff
@@ -1,6 +1,33 @@
-        new GameContext().start();
-    }
+        var wordChoice = new WordSource("/words.txt");
+        wordChoice.readerFile();
```
Метод можно назвать более лаконично, например просто `read`, если хочешь оставить постфикс `File`, то тогда: `readFile`.

А readerFile - переводится как "чтец файла".

```suggestion
        wordChoice.read();
```

shining-mind commented on src/main/java/Application.java:
```diff
@@ -1,6 +1,33 @@
+            }
+
+            System.out.println("Press 'q' to exit");
+            System.out.println("Press 's' to new Game");
```
```suggestion
            System.out.println("Press 's' to start a new game");
```

shining-mind replied:
Но я бы рекомендовал `Press any key`

shining-mind replied:
И лучше сообщение объединить в одно:

`Press 'q' to exit or any key to play a new game`

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,71 @@
+
+public class Game {
+
+    // загаданное слово
```
Рекомендую для документирования использовать формат комментариев javadoc https://java-online.ru/java-javadoc.xhtml

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,71 @@
+
+    // загаданное слово
+    final private String word;
+    private CharSource source;
```
Лучше свойства отбивать отступами друг от друга и оставлять пояснения

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,71 @@
+    public boolean play() {
+        while (this.attemptsHangman.hasDied() && !this.isMatch()) {
+            //собираем сообщение
+            this.answer.printAnswer(this.gallow.getGallow(this.attemptsHangman.getAttempts()) // висилица
```
```suggestion
            this.answer.print(this.gallow.getGallow(this.attemptsHangman.getAttempts()) // висилица
```

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,71 @@
+            // stdin User
+            var userAnswer = this.source.nextChar();
+            // check duplicate
+            if (!this.playerCharList.contains(userAnswer)) {
```
Это нужно в отдельный метод вынести

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,71 @@
+            // check duplicate
+            if (!this.playerCharList.contains(userAnswer)) {
+                var defaultMask = this.maskWord;
+                // check char user for compliance word
```
Комментарии рекомендую начинать с большой буквы: // Сheck entered char for compliace with a word

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,71 @@
+                for (int i = 0; i < this.word.length(); i++) {
+                    if (charWordMatch(userAnswer, this.word, i))
+                        this.maskWord.setCharAt(i, userAnswer);
+                }
```
Между блоками стоит делать отступ

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,71 @@
+            if (!this.playerCharList.contains(userAnswer)) {
+                var defaultMask = this.maskWord;
+                // check char user for compliance word
+                for (int i = 0; i < this.word.length(); i++) {
```
Здесь можно сделать проще:

```java
var index = 0;

while (index > -1) {
  index = this.word.indexOf(char, index);

  if (index != -1) {
    this.maskWord.setCharAt(index, char);
  }
}
```

И это выделить в отдельным метод, например `fillChars(char)`

shining-mind replied:
Ещё будет полезно возвращать булево значение, чтобы сразу знать угадал пользователь  или нет.

shining-mind commented on src/main/java/Source/IntputSource.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.ArrayList;
+
+public interface IntputSource {
+    ArrayList<String> readerFile();
```
Ну, такой интерфейс бесполезен для класса `Game`, ему нужен интерфейс, который возвращает слово.

shining-mind commented on src/main/java/Source/WordSource.java:
```diff
@@ -1,38 +1,47 @@
 
-        try (InputStream is = this.getClass().getResourceAsStream("/words.txt");
+    @Override
+    public ArrayList<String> readerFile() {
```
Такая сигнатура не имеет смысла, так как слова в итоге сохраняются в приватное свойство класса.

shining-mind commented on src/main/java/Source/WordSource.java:
```diff
@@ -1,38 +1,47 @@
             }
         } catch (IOException ex) {
             System.out.println(ex.getMessage());
+            return null;
```
Возвращать null, где ожидается ArrayList - это анти-паттер, тогда уж стоило вернуть пустой массив. Но как правило IO ошибки имеет смысл выбрасывать во вне, потому что если вернуть пустой массив, то может возникнуть иллюзия что данных просто нет в файле.

aleksandrzajbel25182 committed:
Set Google style rules

aleksandrzajbel25182 committed:
Simplify the output of instructions to the user

aleksandrzajbel25182 committed:
Refactor the play method

aleksandrzajbel25182 committed:
Fix a bug reading from a file

aleksandrzajbel25182 committed:
Refactoring file reading

Change the file path.Change the InputSource interface, since the Game class needs an interface that returns the hidden word

aleksandrzajbel25182 committed:
Сreate a "hagman" package

Transfer the AttemptsHangman.java and Hagman.java in the package and connections in the classes used

aleksandrzajbel25182 committed:
Reformat code

aleksandrzajbel25182 committed:
Install the javadoc plugin

aleksandrzajbel25182 committed:
Documenting the code

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+
+      var userAnswer = nextChar();
+      if (!checkDuplicate(userAnswer)) {
+        continue;
```
Старайся не использовать continue и break без лишней необходимости

shining-mind replied:
Почему?

Если тело цикла большое, то очень легко запутаться где прерывается его работа, потому рекомендуется использовать ветвление.

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+      this.answer.print(message);
+
+      var userAnswer = nextChar();
+      if (!checkDuplicate(userAnswer)) {
```
checkDuplicate - непонятно что он именно проверяет, что это дубль или что не дубль, потому лучше использовать другую форму записи `isDuplicate`, тогда очевидно что именно проверяется

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+        continue;
+      }
+
+      if (!checkCharWord(userAnswer)) {
```
здесь аналогичное замечание

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+
+      if (!checkCharWord(userAnswer)) {
+
+        this.attemptsHangman.registerAttempts();
```
может `registerAttempt`? у нас ведь попытка регистрируется, а не попытки

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+   */
+  private Gallows gallow;
+
+  private ArrayList<Character> playerCharList = new ArrayList<>();
```
Здесь лучше подойдет структура данных `Set`, потому что поиск вхождения в множестве `O(1)`, в массиве же `O(n)` 

Про нотацию большого О https://tproger.ru/translations/data-structure-time-complexity-in-python

shining-mind replied:
Также стоит переименовать эту переменную, например: `enteredChars`/`usedChars`/`guessedChars`. 

В общем, чтобы по смыслу было понятно, что тут хранятся введенные символы.

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+      if (!checkCharWord(userAnswer)) {
+
+        this.attemptsHangman.registerAttempts();
+        this.playerCharList.add(userAnswer);
```
операция добавления символа в список использованных общая, потому можно вынести её за пределы if/else

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+  /**
+   * The method of checking a character in a word
+   *
+   * @param ch Проверяемый символ
```
Не смешивать разные языки: либо английский, либо русский

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+
+    var index = 0;
+    boolean flag = false;
+    //Нахождение символа в слове и измениние маски
```
Стоит всегда ставить пробел после косой черты
```suggestion
    // Нахождение символа в слове и измениние маски
```

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+  private boolean checkCharWord(char ch) {
+
+    var index = 0;
+    boolean flag = false;
```
лучше дать семантичное название переменной, например: `hasChar`/`isInWord`/`matches`

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+    if (!this.playerCharList.contains(ch)) {
+      return true;
+    } else {
+      this.answer.print("Вы уже вводили такую букву");
```
Это пример сайд-эффекта. Почему так делать не стоит: нам может потребоваться проверить некоторое предусловие N раз, либо выполнить какие-то расчёты. При этом будет странно если при повторном выполнении расчёта мы будет в stdout что-то писать.

В некоторых случаях методы могут быть просто декомпозицией основного алгоритма, тогда мы можем что-то писать в stdout, но в данном случае это не тот кейс.

shining-mind commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,160 @@
+   * @return The next character
+   * @see CharSource#nexChar()
+   */
+  private Character nextChar() {
```
А точно ли есть смысл в этом методе? По-моему нормально использовать в коде `this.source.nextChar` - это как раз пример композиции в ООП (https://habr.com/ru/articles/354046/). Как правило для композитов (в нашем случае source) не имеет смысла создавать обертки в родительском классе.

shining-mind commented on src/main/java/hagman/AttemptsHangman.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * AttemptsHangman class that implements the interface {@link Hagman}
+ */
+public class AttemptsHangman implements Hagman {
```
Чем больше смотрю на этот класс, тем больше кажется, что он лишний. Всё-таки можно попытки хранить в самой игре. 

Этот класс имел бы смысл, если бы он генерировал кастомную виселицу в зависимости от количества попыток. А так представление виселицы хранится в Gallows, который не связан с этим классом.

aleksandrzajbel25182 committed:
Fixed a typo in the documentation

aleksandrzajbel25182 committed:
Implement attempts in the Game class

Delete the Hagman package.
 Attempts and actions on them are better stored in the Game class itself

aleksandrzajbel25182 committed:
Fix error

aleksandrzajbel25182 committed:
Change ArrayList to HasSet

The expected execution time of the HashSet.contains() function is O(1)
For comparison, ArrayList.contains() = O(n)
Change the name of "plaeyrCharList" to "usedChar"

aleksandrzajbel25182 committed:
Remove the nextChar method

aleksandrzajbel25182 committed:
Rename the method

the checkDuplicate method is a predicate, add a prefix "is"

aleksandrzajbel25182 committed:
Rename the method

aleksandrzajbel25182 committed:
Code refactoring

aleksandrzajbel25182 committed:
Fix bug and polishing
